### PR TITLE
obs-ffmpeg: NVENC additional settings.

### PR DESCRIPTION
**Pull request by Dusre**
Added missing features of nvenc.

### CODE REVIEW

**Comment by Dusre** [comment]
Quote from the latest NVENC_VideoEncoder_API_ProgGuide.pdf:

> Game-casting &
cloud transcoding
➢ High quality preset
➢ Rate control mode = CBR
➢ Medium VBV buffer size (1 second)
➢ B Frames*
➢ Look-ahead
➢ B frame as reference
➢ Finite GOP length (2 seconds)
➢ Adaptive quantization (AQ) enabled**
*: Recommended for low motion games and natural video.
**: Recommended on second generation Maxwell GPUs and above. 

So i dont get why these would be recommended by nvidia them self if its bad.

**Comment by Fenrirthviti** [comment]
Sorry, maybe I wasn't quite clear on what I meant. Currently, you have the options translated as:

```NVENC.noscenecut="No-Scenecut (When lookahead is enabled, enable this)"
NVENC.badapt="B-Adapt (When lookahead is enabled, disable this)"```

This shouldn't be left to the user, IMO, if it's not recommended and can cause issues. There should be a programmatic check to enable/disable, or at the very least, a warning sent to the UI instead of putting this in the string of the option itself.

**Comment by h1z1** [comment]
Interesting.  Only version of this guide has no mention of the text above?  Kind of odd they'd recommend CBR.  Not saying it doesn't exist but it's not clear what version is being referenced.

**Comment by Dusre** [comment]
Why would you use VBR for streaming? You can get the documents from here https://developer.nvidia.com/designworks/video_codec_sdk/downloads/v8.1

**Comment by h1z1** [comment]
Because it's technically better?  CBR breaks down as you run out of bitrate bandwidth.  Sites like Twitch recommend it because it helps THEM control traffic bursts and I don't think their old flash player handled ABR too well.    Given that, should settings like this not be part of their specific service config?

Your link requires a login.

**Comment by RytoEX** [comment]
@Dusre
Before I dig into this PR, is [this thread](https://obsproject.com/forum/threads/nvenc-adaptive-quantization.65343/) also you?  I'd just like to establish some background information first.


@h1z1
The newer versions of the NVIDIA SDK documentation and developer docs require an NVIDIA Developer Program account, which is a free signup (but yes, it is a required signup).  Some of the older versions listed [here](https://developer.nvidia.com/video-codec-sdk-archive) do not require an account, but they obviously will not have information on newer NVIDIA features.

As for why NVIDIA recommends CBR for "Game-casting & cloud transcoding", I'd guess it's exactly because it helps transcoders (or services that transcode) handle traffic bursts better.  To really know why NVIDIA made that recommendation, you'd have to ask NVIDIA.

**Comment by Dusre** [comment]
Yes thats my post but i now decided to post here since no one was interested there.
Also VBR is configured wrong in the code example when you put 6k bitrate it can burst even 10k and then you gonna see dropped frames.
This example fixes that problem so it doesnt go over the max bitrate.

> 	else if (astrcmpi(rc, "cbr") == 0) { /* CBR by default */
		av_opt_set_int(enc->context->priv_data, "cbr", true, 0);
		enc->context->rc_max_rate = bitrate * 1000;
		enc->context->rc_min_rate = bitrate * 1000;
		cqp = 0;
	}
	else if (astrcmpi(rc, "vbr") == 0) {
		enc->context->rc_max_rate = bitrate * 1000;
		cqp = 0;
	}

**Comment by RytoEX** [comment]
Okay then.  It's not that there was no interested there, it's just that the video encoder flags/options can be very technical, and not everyone has the technical knowledge to be able to thoroughly understand them and how they should be implemented, myself included.

You're working off of older FFmpeg default settings, so we're going to have to bring some of that up to date.  I'll try to cover some of that in an upcoming code review.

File: plugins/obs-ffmpeg/data/locale/en-US.ini [hunk 1]
```diff
@@ -18,6 +18,12 @@ NVENC.Preset.ll="Low-Latency"
 NVENC.Preset.llhq="Low-Latency High Quality"
 NVENC.Preset.llhp="Low-Latency High Performance"
 NVENC.Level="Level"
+NVENC.SpatialStrength="Spatial AQ Strength"
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/data/locale/en-US.ini]: You should probably use "NVENC.SpatialAQ.Strength" for the string name and place it below "NVENC.SpatialAQ".

File: plugins/obs-ffmpeg/data/locale/en-US.ini [hunk 2]
```diff
@@ -18,6 +18,12 @@ NVENC.Preset.ll="Low-Latency"
...
+NVENC.rclookahead="RC-Lookahead"
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/data/locale/en-US.ini]: You should probably use "NVENC.RCLookahead" here.

File: plugins/obs-ffmpeg/data/locale/en-US.ini [hunk 3]
```diff
@@ -18,6 +18,12 @@ NVENC.Preset.ll="Low-Latency"
...
+NVENC.surfaces="Surfaces"
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/data/locale/en-US.ini]: Use "NVENC.Surfaces" here.

File: plugins/obs-ffmpeg/data/locale/en-US.ini [hunk 4]
```diff
@@ -18,6 +18,12 @@ NVENC.Preset.ll="Low-Latency"
...
+NVENC.SpatialAQ="Use Spatial AQ"
+NVENC.TemporalAQ="Use Temporal AQ"
+NVENC.forcedidr="Forced-IDR"
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/data/locale/en-US.ini]: Use "NVENC.ForcedIDR" here.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 5]
```diff
@@ -139,6 +139,12 @@ static bool nvenc_update(void *data, obs_data_t *settings)
 	int gpu = (int)obs_data_get_int(settings, "gpu");
 	bool cbr_override = obs_data_get_bool(settings, "cbr");
 	int bf = (int)obs_data_get_int(settings, "bf");
+	int spstrength = (int)obs_data_get_int(settings, "spatialaqstrength");
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: If you want to abbreviate "Spatial AQ Strength" for the variable here, perhaps "saqstrength" would be slightly more appropriate?

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 6]
```diff
@@ -183,11 +189,22 @@ static bool nvenc_update(void *data, obs_data_t *settings)
 		enc->context->rc_min_rate = bitrate * 1000;
 		cqp = 0;
 	}
-
+
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: You added a tab character here.  Please remove it.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 7]
```diff
@@ -183,11 +189,22 @@ static bool nvenc_update(void *data, obs_data_t *settings)
 		enc->context->rc_min_rate = bitrate * 1000;
 		cqp = 0;
 	}
-
+	
+	if (rclookahead > 0)
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style:  This `if` should have its opening curly brace at the end of this line.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 8]
```diff
@@ -210,24 +227,33 @@ static bool nvenc_update(void *data, obs_data_t *settings)
...
-	     "\t2-pass:       %s\n"
-	     "\tb-frames:     %d\n"
-	     "\tGPU:          %d\n",
-	     rc, bitrate, cqp, enc->context->gop_size,
-	     preset, profile, level,
-	     enc->context->width, enc->context->height,
-	     twopass ? "true" : "false",
-	     enc->context->max_b_frames,
-	     gpu);
+		"\trate_control: %s\n"
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: You changed the whitespace in this section.  Please revert your whitespace changes and present only the functional changes.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 9]
```diff
@@ -183,11 +189,22 @@ static bool nvenc_update(void *data, obs_data_t *settings)
 		enc->context->rc_min_rate = bitrate * 1000;
 		cqp = 0;
 	}
-
+	
+	if (rclookahead > 0)
+	{
+		av_opt_set_int(enc->context->priv_data, "b_adapt", false, 0);
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: You aren't exposing these options to the user.  They're just being forced to their non-default settings.  Why?

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c (hunk 7 above)
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Instead of adding this `if` block here, I think it would be better placed below the block of `av_opt_set_int` calls below.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 10]
```diff
@@ -402,6 +428,12 @@ static void nvenc_defaults(obs_data_t *settings)
 	obs_data_set_default_bool(settings, "2pass", true);
 	obs_data_set_default_int(settings, "gpu", 0);
 	obs_data_set_default_int(settings, "bf", 2);
+	obs_data_set_default_bool(settings, "spatialaq", false);
+	obs_data_set_default_bool(settings, "temporalaq", false);
+	obs_data_set_default_bool(settings, "forcedidr", false);
+	obs_data_set_default_int(settings, "spatialaqstrength", 1);
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: The default setting for `aq-strength` (spatialaqstrength) should be 8, according to FFmpeg.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 11]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
 	obs_properties_add_int(props, "gpu", obs_module_text("GPU"), 0, 8, 1);
 
 	obs_properties_add_int(props, "bf", obs_module_text("BFrames"),
-			0, 4, 1);
+		0, 4, 1);
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: You changed the whitespace here.  Please revert this change.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 12]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+	obs_properties_add_bool(props, "spatialaq",
+		obs_module_text("NVENC.SpatialAQ"));
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style:  Tab in twice to indicate a line continuation.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 13]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+	obs_properties_add_bool(props, "temporalaq",
+		obs_module_text("NVENC.TemporalAQ"));
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style:  Tab in twice to indicate a line continuation.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 14]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+	obs_properties_add_bool(props, "forcedidr",
+		obs_module_text("NVENC.forcedidr"));
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style:  Tab in twice to indicate a line continuation.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 15]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+	obs_properties_add_int(props, "spatialaqstrength",
+		obs_module_text("NVENC.SpatialStrength"),
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style:  Tab in twice to indicate a line continuation.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 16]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+		1, 15, 1);
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style:  This line can safely be at the end of the preceding line without exceeding the column limit.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 17]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+	obs_properties_add_int(props, "rclookahead",
+		obs_module_text("NVENC.rclookahead"),
+		0, 32, 1);
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style:  This line can safely be at the end of the preceding line without exceeding the column limit.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 18]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+	obs_properties_add_int(props, "surfaces",
+		obs_module_text("NVENC.surfaces"),
+		32, 64, 1);
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style:  This line can safely be at the end of the preceding line without exceeding the column limit.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 19]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+		obs_module_text("NVENC.SpatialAQ"));
+	obs_properties_add_bool(props, "temporalaq",
+		obs_module_text("NVENC.TemporalAQ"));
+	obs_properties_add_bool(props, "forcedidr",
+		obs_module_text("NVENC.forcedidr"));
+	obs_properties_add_int(props, "spatialaqstrength",
+		obs_module_text("NVENC.SpatialStrength"),
+		1, 15, 1);
+	obs_properties_add_int(props, "rclookahead",
+		obs_module_text("NVENC.rclookahead"),
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style: Tab in twice to indicate a line continuation.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 20]
```diff
@@ -512,7 +544,22 @@ static obs_properties_t *nvenc_properties(void *unused)
...
+		0, 32, 1);
+	obs_properties_add_int(props, "surfaces",
+		obs_module_text("NVENC.surfaces"),
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: Style: Tab in twice to indicate a line continuation.

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c (hunk 18 above)
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: The range for `surfaces` should be an integer from 0 to 64.  These three numbers should be `0, 64, 1`

File: plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c [hunk 21]
```diff
@@ -402,6 +428,12 @@ static void nvenc_defaults(obs_data_t *settings)
 	obs_data_set_default_bool(settings, "2pass", true);
 	obs_data_set_default_int(settings, "gpu", 0);
 	obs_data_set_default_int(settings, "bf", 2);
+	obs_data_set_default_bool(settings, "spatialaq", false);
+	obs_data_set_default_bool(settings, "temporalaq", false);
+	obs_data_set_default_bool(settings, "forcedidr", false);
+	obs_data_set_default_int(settings, "spatialaqstrength", 1);
+	obs_data_set_default_int(settings, "rclookahead", 0);
+	obs_data_set_default_int(settings, "surfaces", 32);
```
Comment (RytoEX) [inline plugins/obs-ffmpeg/obs-ffmpeg-nvenc.c]: The default setting for `surfaces` should be 0, according to FFmpeg.

**Comment by jp9000** [comment]
Hi there, thank you for the PR, however I have some disagreement with adding these specific settings directly as OBS settings.  Instead, I would prefer an option similar to how x264 works for advanced settings (its "x264opts" setting), where you can use a settings string to specify additional settings that are available.  I realize that this may be a bit frustrating, but I would prefer to respectfully reject it in favor of using a settings string instead.

Additionally, there are a few important notes about your pull request:

1.) You didn't follow the contribution guidelines.  Please see RytoEX's review comments, and read the "Coding guidelines" and "Commit guidelines" sections of https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst
2.) You have multiple duplicate commits, and an unnecessary merge commit.  This is usually an indicator of lack of git experience.  I'm assuming you performed a merge rather than a rebase.  You should have rebased to fix the commits of your pull request and then force pushed to your fork's branch associated with this pull request (although in this case you used your fork's master branch, which you typically shouldn't do with a pull request).  I would also recommend taking the time to review our commit history, as that can give you a pretty quick understanding of how our commits and code normally look.

I will be willing to allow the additional NVENC settings, however, I definitely have to opt for an options string rather than have additional advanced parameters as properties.

I realize that having proposed changes can be frustrating -- please don't let that discourage you from submitting pull requests in the future.
